src/framework/input/element-input.js

```diff
@@ -659,33 +682,44 @@ class ElementInput {
         if (!this._enabled) return;
 
         const newTouchedElements = this._determineTouchedElements(event);
-
+        const { _touchedElements, _touchesForWhichTouchLeaveHasFired } = this;
         for (let i = 0, len = event.changedTouches.length; i < len; i++) {
             const touch = event.changedTouches[i];
-            const newTouchInfo = newTouchedElements[touch.identifier];
-            const oldTouchInfo = this._touchedElements[touch.identifier];
-
+            const { identifier } = touch;
+            const newTouchInfo = newTouchedElements[identifier];
+            const oldTouchInfo = _touchedElements[identifier];
             if (oldTouchInfo) {
-                const coords = this._calcTouchCoords(touch);
-
+                const { x, y } = this._calcTouchCoords(touch);
+                const leaveHasFired = _touchesForWhichTouchLeaveHasFired[identifier];
+                const leftPreviousElement  = !newTouchInfo || newTouchInfo.element !== oldTouchInfo.element;
+                const enterPreviousElement =  newTouchInfo && newTouchInfo.element === oldTouchInfo.element;
                 // Fire touchleave if we've left the previously touched element
-                if ((!newTouchInfo || newTouchInfo.element !== oldTouchInfo.element) && !this._touchesForWhichTouchLeaveHasFired[touch.identifier]) {
-                    this._fireEvent('touchleave', new ElementTouchEvent(event, oldTouchInfo.element, oldTouchInfo.camera, coords.x, coords.y, touch));
-
+                if (leftPreviousElement && !leaveHasFired) {
+                    this._fireEvent('touchleave', new ElementTouchEvent(event, oldTouchInfo.element, oldTouchInfo.camera, x, y, touch));
                     // Flag that touchleave has been fired for this touch, so that we don't
                     // re-fire it on the next touchmove. This is required because touchmove
                     // events keep on firing for the same element until the touch ends, even
                     // if the touch position moves away from the element. Touchleave, on the
                     // other hand, should fire once when the touch position moves away from
                     // the element and then not re-fire again within the same touch session.
-                    this._touchesForWhichTouchLeaveHasFired[touch.identifier] = true;
+                    _touchesForWhichTouchLeaveHasFired[identifier] = true;
                 }
-
-                this._fireEvent('touchmove', new ElementTouchEvent(event, oldTouchInfo.element, oldTouchInfo.camera, coords.x, coords.y, touch));
+                // And fire touchenter if we've entered the previously left touched element again
```

Comment:
I feel like we should fire touchenter if a user has clicked outside the element and moved into it as well. Seems odd to only fire it if we started in an element, leave and then re-enter